### DISCUSSION TITLE
Conditional skip of property assignment in Object Initializer

### DISCUSSION THREAD

**Discussion by GabrielHalvonik** [dotnet/csharplang#7413] [Language Ideas]
Hi,
I've encountered cases where it might be useful to enhance conditional property assignment inside Object Initializer and would like to discuss this approach.
In this approach it will be possible to perform property initialization if condition is met, otherwise the initialization process will be skipped (not just assigned to default value). 
My main desire for this feature come from UI programming with nested Object Initializers, where workaround for this approach is unpleasant. Especially when performing Collection Initialization inside Object Initializer, where it is not possible, to skip object addition if condition is not met (adding default value could easily break underlying functionality of the collection...e.g. when adding default value for referenceable type and collection does not count with null values).
Example of this approach:

```csharp 
class Control { }
class ScrollBar : Control { }
class Decorator : Control {
    public Control? Child { get; set; }
}

class Panel : Control {
    public List<Control> Children { get; set; } = new ();
}

class ScrollArea {

    public Control? Content { get; set; }
    public bool EnableScrollBar { get; set; } = false;

    void ConstructControl() {
        Content = new Decorator {
            Child = new Panel {
                Children = {
		    new Control(),
		    EnableScrollBar ? new ScrollBar() : none,
	            new Control(),
		}
             }
	 };
    }

}
```

(notice this line: `EnableScrollBar ? new ScrollBar() : none`)

Workaround for this approach will be to reference whole object initialization chain until stumble upon object owning that collection property and subsequently perform conditional list addition, which is quite lengthy and not very elegant way. It also can make it more difficult to do refactoring and need to keep in mind these reference chains. 

Also possible workaround is to create extension method for initialization and skip the necessity of doing reference chaining.
sample:
```csharp 
public static class Extensionss {
    public static T Init<T>(this T self, Action<T> action) where T : class {
         action(self);
         return self;
    }
}

...

Content = new Decorator {
    Child = new Panel {
        Children = {
            new Control(),
            new Control(),
        }
    }.Init(self => {
        if (condition) self.Children.Insert(1, new ScrollBar()); 
    })
};
```

I can imagine it will be possible to convert ```(condition) ? value : none ``` into something like code above workaround under-the hood without much of hassle, but I will keep such a conclusion for someone more competent in tis area.

Also it may be convenient to not to restrict such a ability only in Collection Initializer, but also in Object Initializer. Even though assignment  of ```default``` value is not problem as it is in Collection Initializer, but it can serve pretty handy in case of keeping default property value untouched if assignment condition is not met.
e.g:
```csharp 
class Control {
	public IBrush? Background { get; set; } = Brushes.DarkGray;
}

class Window {

    public (bool Dark, bool White) Theme { get; set; }
    public Control? Child { get; set; }

    void ConstructControl() {
        Child = new Control {
	    Background = Theme switch {
	        { Dark: true } => Brushes.Black,
		{ White: true } => Brushes.White,
		_ => none
	    }
        };
    }
}
```

Please, keep in mind, that those samples are just for clear illustration of the feature. UI programming is especially common for having multiple nesting with possibly conditional assignments, but I find this feature to be useful also outside of scope of UI programming... Especially inside Collection Initializers. 
I I would love to hear your thoughts on this feature.

**Comment by HaloFour** [comment]
See also: #5588

**Comment by GabrielHalvonik** [comment]
> See also: #5588

Thanks, missed that one. Seems like there was chance of consideration to materialise that feature at some point. 
Personally from that discussion, I find:
```csharp 
new A
{
    B = b when readB,
    C = c
};
```
with `when` keyword to be pretty neat and non-invasive with new keywords, but in case of multiple conditions, approach mentioned in OP seems to be more pleasant to me.
Maybe if it was possible to chain those conditions this way, it will be more consistent (without necessity of introduction new keyword (`none`)) 
e.g:
```csharp
new Control {
    Background = Brushes.Black when Theme is {  Dark: true } or Brushes.White when Theme is {  White: true }
}
```	 

Maybe it will be welcomed by more devs to bring this feature back on the table. Would like to hear some more specific view from language designers on this.

**Comment by TahirAhmadov** [comment]
See this comment: https://github.com/dotnet/csharplang/issues/5354#issuecomment-1439287957
Basically, if support for collection literals is added to object initializers, then you can do:
```cs

    void ConstructControl() {
        Content = new Decorator {
            Child = new Panel {
                Children += [ // note the syntax I proposed for "adding a collection"
		    new Control(),
		    EnableScrollBar ? ..[new ScrollBar()] : ..[], // and this is already draft spec
	            new Control(),
		]
             }
	 };
    }
```

**Reply by GabrielHalvonik** [reply]
Yep, played little bit with that.

Anyway having solid feature for this (and also for object initialization properties) will IMO be more consistent and will not feel like weird walk-around, that feels more like side-effect then targeted approach. However better than nothing.
